core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -95,6 +95,8 @@ class KafkaController(val config: KafkaConfig,
     new ControllerBrokerRequestBatch(config, controllerChannelManager, eventManager, controllerContext, stateChangeLogger))
   val topicDeletionManager = new TopicDeletionManager(config, controllerContext, replicaStateMachine,
     partitionStateMachine, new ControllerDeletionClient(this, zkClient))
+  val reassignmentsManager = new ReassignmentsManager(controllerContext, zkClient, topicDeletionManager,
```

Comment:
High level, I think this refactor is a good improvement. It is similar to the refactor that introduced `TopicDeletionManager` and it does make testing a bit easier. That said, I want to mention a kind of drawback. Although it succeeds in factoring out some of the logic out of `KafkaController`, it doesn't really do anything about the complex interdependencies between the various components and the fact that any one of them can mutate the controller state. And in fact, it makes it a little harder to track all these mutations because they are spread over more classes.

I think it would be a useful exercise to try and think about some of these components in more of a functional way. Rather than allowing the reassignment manager to directly mutate any and all state that the controller owns, perhaps we can treat it more like a function which accepts the current state of the world, makes some modifications, and then returns the new state of the world. Then it could be up to the controller to decide how to enact the new state (e.g. by making changes in ZK and sending UpdateMetadata requests). The nice thing then is that we don't need all the dependencies and all the nasty mocking that comes with them.

Anyway, this is more of a "food for thought" comment. I'm not exactly sure how to do this myself.

Reply:
I do agree it does not fix the underlying issue of complexity and I agree it makes the mutation changes harder to track.

If we are to start refactoring towards a more functional approach, I think it would be easier to start with the lowest level classes which still mutate the state - the state machines.
Otherwise it'd be pretty difficult to make the ReassignmentManager functional when components it uses mutate the state underneath